### PR TITLE
[UI] Restore Registry tree scroll position when returning to a tab

### LINKED ISSUE #21864: [UI] Registry tree loses scroll position when switching back to a previously scrolled tab
Scroll down the Registry Models tree, switch to another registry tab (for example, Relationships), and then switch back to Models. The tree is reset to the top instead of preserving the previous scroll position.

The component already contains logic to save and restore the scroll position, but the save condition is incorrect, so the restoration logic never gets a saved offset.

### Cause

`ui/components/registry/MesheryTreeView.tsx` guards the scroll offset with:

```tsx
if (!data.length === 0) {
  scrollRef.current = div.scrollTop;
}
```

This is evaluated as `(!data.length) === 0`, comparing a boolean with a number. The condition is therefore never true, so `scrollRef.current` is never assigned. Since the restore effect is also gated on that ref, it never restores the previous scroll position.

The restore logic also uses `document.querySelector('.scrollElement')` to find the scroll container. This element is rendered conditionally, since the empty and no-result states replace it. If the list becomes empty, the query can return `null`. Using a ref for the container ensures the restoration targets this component's own scroll element and safely skips restoration when the container is not rendered.

The saved offset is also checked with a truthiness condition. Since `0` is a valid scroll position, the check should explicitly compare against `null` so that a saved position of `0` is not treated as missing.

### Measured

Local provider with 584 models registered, Registry → Models:

| Step                      |             scrollTop |
| ------------------------- | --------------------: |
| Scrolled down in Models   |                  2946 |
| Switched to Relationships |                     0 |
| Switched back to Models   | **0 — position lost** |

### Expected Behavior

The Registry tree should preserve and restore the user's previous scroll position when switching between registry tabs.

For example, if the Models tree is scrolled to `2946`, switching to another tab and then returning to Models should restore the tree to approximately `2946` instead of resetting it to the top.

The restoration logic should also safely handle cases where the scroll container is not rendered.

### To Reproduce

1. Open Registry → Models.
2. Scroll down the Models tree.
3. Switch to the Relationships tab.
4. Switch back to the Models tab.
5. Observe that the Models tree has returned to the top instead of the previous scroll position.

### Screenshots/Logs

Not applicable.

### Environment

* Browser: Chrome
* Host OS: macOS
* Meshery Server Version: stable-v
* Platform: Docker

---

### Contributor [[[Guides](https://docs.meshery.io/project/contributing)](https://docs.meshery.io/project/contributing)](https://docs.meshery.io/project/contributing) and [[[Handbook](https://meshery.io/community#handbook)](https://meshery.io/community#handbook)](https://meshery.io/community#handbook)

* 🎨 Wireframes and [[[designs for Meshery UI](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI) in Figma [[[(open invite)](https://www.figma.com/team_invite/redeem/GvB8SudhEOoq3JOvoLaoMs)](https://www.figma.com/team_invite/redeem/GvB8SudhEOoq3JOvoLaoMs)](https://www.figma.com/team_invite/redeem/GvB8SudhEOoq3JOvoLaoMs)
* 🖥 [[[Contributing to Meshery UI](https://docs.meshery.io/project/contributing/contributing-ui)](https://docs.meshery.io/project/contributing/contributing-ui)](https://docs.meshery.io/project/contributing/contributing-ui)
* 🙋🏾🙋🏼 Questions: [[[Discussion Forum](https://discuss.meshery.io/)](https://discuss.meshery.io//)](https://discuss.meshery.io)
* 📺 [[[Self-paced Contributor Trainings](https://meshery.io/talks-and-trainings#trainings)](https://meshery.io/talks-and-trainings#trainings)](https://meshery.io/talks-and-trainings#trainings)

I would like to contribute and will raise a PR soon. Thank you! :)

### PR DESCRIPTION
**Pull request by Somilg11**
* This PR fixes #21864

`ui/components/registry/MesheryTreeView.tsx` saved the scroll offset behind `if (!data.length === 0)`. This parses as `(!data.length) === 0`, comparing a boolean against a number. The condition was never true, so `scrollRef.current` was never assigned, and the restore effect, which is gated on that same ref, never ran. Both parts of the scroll save/restore logic have effectively been dead since `764fc416c99`.

This is visible when switching between registry tabs. Switching tabs unmounts and remounts the scrollable container, so returning to a tab that was previously scrolled resets the list to the top.

**Three changes:**

1. Compare the length directly (`data.length !== 0`) so the current scroll offset is recorded when the list contains data.
2. Hold the scroll container in a ref instead of using `document.querySelector('.scrollElement')`. The element is conditionally rendered, so the query can return `null` when the list is empty. This was harmless while the save logic was broken because the restore effect never ran, but would cause a null dereference once restoration started working. Using a ref also ensures the restore targets this component's own scroll container instead of the first matching element in the document.
3. Check the saved offset with `!== undefined` instead of a truthiness check, so a valid saved offset of `0` is restored correctly.
4. Restore runs in useLayoutEffect so the offset is applied before paint. Save and restore are both keyed by view.

**Before / after** (local server, 584 models, Registry → Models → Relationships → Models):

|                | before | after   |
| -------------- | ------ | ------- |
| scrolled down  | 2946   | 3018    |
| back on Models | **0**  | **146** |

The `146` after the fix is expected. When the Models tab is mounted again, only the first page of records is initially loaded. The saved offset is restored, but the shorter list cannot scroll that far, so the browser clamps `scrollTop` to the bottom of the currently loaded content. As more records are loaded, the saved position can be restored further.

The important difference is that the scroll position is no longer reset to `0` on returning to the tab.

Videos attached: before (position lost) and after (position restored).

**Testing**

Two cases were added to `MesheryTreeView.test.tsx`. The first is the regression test for the original issue, and was verified to fail against the unfixed component:

Before:

```text
× restores the scroll offset after more records arrive
  AssertionError: expected +0 to be 120 // Object.is equality
  Tests  1 failed | 12 passed (13)
```

https://github.com/user-attachments/assets/79d32bd2-a05b-4284-bd30-9c2a732cf5e9

After:

```text
Test Files  1 passed (1)
     Tests  13 passed (13)
```

https://github.com/user-attachments/assets/24692a88-52e7-404a-985f-76bfcba9a293

Only the first test directly proves the original defect. The second test verifies that emptying the list while an offset is saved does not throw. It also passes on the old code because the restore path was unreachable there; it guards the new container-ref logic from introducing a crash.

`eslint`, `prettier --check` and `tsc --noEmit` are clean for this file.

**[[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**

* [x] Yes, I signed my commits.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved scroll-position restoration when loading additional registry records.
  * Preserved scroll positions during list updates, including when the saved position is at the top.
  * Preserved each registry view’s independent scroll position when switching between views.
  * Prevented errors when the registry transitions to an empty “No result found” state.
  * Ensured returning to a previously viewed registry section restores its last scroll position.

### CODE REVIEW

**Review by dhruveshmishra — COMMENTED** [review]
Address ai suggestions and Use useLayoutEffect instead of useEffect when restoring scrollTop so the scroll position is set before the browser paints, preventing a visible jump.

File: ui/components/registry/MesheryTreeView.tsx [hunk 1]
```diff
@@ -95,15 +101,21 @@ const MesheryTreeView = React.memo(
...
 
     useEffect(() => {
-      if (scrollRef.current) {
-        const div = document.querySelector('.scrollElement');
-        div.scrollTop = scrollRef.current;
+      const container = scrollContainerRef.current;
+      // Compare against null explicitly: an offset of 0 is a legitimate saved
+      // position, and a truthiness check would silently skip restoring it.
+      if (container && scrollRef.current !== null) {
+        container.scrollTop = scrollRef.current;
```
[the remark this answers is not in the document]
Comment (Somilg11) [reply ui/components/registry/MesheryTreeView.tsx]: Fixed. Offsets are keyed by view now, and a Models → Relationships → Models test asserts each view restores its own offset rather than the last one scrolled. It fails against a single shared offset with expected 40 to be 120.

File: ui/components/registry/MesheryTreeView.tsx [hunk 2]
```diff
@@ -95,17 +105,27 @@ const MesheryTreeView = React.memo(
...
-      if (scrollRef.current) {
-        const div = document.querySelector('.scrollElement');
-        div.scrollTop = scrollRef.current;
+      const container = scrollContainerRef.current;
+      const savedOffset = scrollOffsetsRef.current[view];
+      // Compare against undefined explicitly: an offset of 0 is a legitimate
+      // saved position, and a truthiness check would silently skip restoring it.
+      if (container && savedOffset !== undefined) {
+        container.scrollTop = savedOffset;
       }
```
[the remark this answers is not in the document]
Comment (Somilg11) [reply ui/components/registry/MesheryTreeView.tsx:133]: Fixed. Offsets are keyed by view now, and a Models → Relationships → Models test asserts each view restores its own offset rather than the last one scrolled. It fails against a single shared offset with expected 40 to be 120.

File: ui/components/registry/MesheryTreeView.tsx [hunk 3]
```diff
@@ -95,17 +105,27 @@ const MesheryTreeView = React.memo(
...
 
     useEffect(() => {
-      if (scrollRef.current) {
-        const div = document.querySelector('.scrollElement');
-        div.scrollTop = scrollRef.current;
+      const container = scrollContainerRef.current;
+      const savedOffset = scrollOffsetsRef.current[view];
+      // Compare against undefined explicitly: an offset of 0 is a legitimate
+      // saved position, and a truthiness check would silently skip restoring it.
+      if (container && savedOffset !== undefined) {
```
[the remark this answers is not in the document]
Comment (Somilg11) [reply ui/components/registry/MesheryTreeView.tsx:131]: You're right, and the description was the stale half. It described an earlier version that stored a single number | null; once offsets moved into a Record<string, number> a missing key reads as undefined, so !== undefined is the correct check. I've corrected the PR description and added a comment in the code so the choice is obvious to the next reader.

Copilot also flagged separately that nothing tested the case this comparison exists for — every test saved a non-zero offset. There's now a test that saves 0, moves the container, re-renders and asserts it returns to 0. It fails against a truthiness check.

**Comment by Somilg11** [comment]
@dhruveshmishra Thanks — both done in 319438df003.
The restore now runs in useLayoutEffect, so the offset is applied before the browser paints rather than painting at the top and then jumping. Good call; that's the kind of thing that resurfaces later as a "flicker" bug nobody can pin down.

On the AI suggestions: CodeRabbit's per-view offsets, and Copilot's key-consistency, null/undefined, offset-0 coverage and empty-state remount points are all addressed — replied on each thread.

Also rebased onto current master, so the branch is a single commit again instead of carrying a merge commit.

**Review by KumarNirupam1 — COMMENTED** [review]
Tested locally and confirmed the fix — scroll position in Models survives switching to another tab and back now. I pushed each tab to a different scroll spot to check they were independent and they were. Empty-state and top-of-list both behave. Nice catch on that !data.length === 0 line, that was never saving anything. 

lgtm 👍

https://github.com/user-attachments/assets/36c31534-37a0-414c-aa70-f1730cbb4fff

**Comment by ritzorama** [comment]
Great focused fix. Per-view offset storage, ref-scoped restoration, and the targeted zero/remount/tab-switch tests make this regression coverage notably stronger. Thanks for following through on the earlier review feedback.
